### PR TITLE
Changes to support cloud based Redis instances

### LINKED ISSUE #46: Support for cloud based Redis instance
**Problem**
Currently, the set up instructions for the RedisStreamSource have us install a cluster local version of redis. We should be able to use other instances of Redis as well, for example a cloud provided Redis instance.

**[Persona:](https://github.com/knative/eventing/blob/master/docs/personas.md)**
Which persona is this feature for?
Event Producer & Consumer

**Exit Criteria**
A measurable (binary) test that would indicate that the problem has been resolved.
I am able to use a redis instance from a cloud provider instead of the locally installed sample/redis.

**Time Estimate (optional):**
How many developer-days do you think this may take to resolve?

**Additional context (optional)**
Add any other context about the feature request here.

### PR DESCRIPTION
**Pull request by aavarghese**
Signed-off-by: Ansu Varghese <avarghese@us.ibm.com>

Fixes #46 

## Proposed Changes

- Redis Connection code now supports non-local Redis instance too (for ex, a Redis instance from IBM Cloud). `Address` field in RedisStreamSource spec now requires a full connection string, and a TLS Certificate passed in via configmap. (Local Redis instances will continue to work)

**Release Note**

```
RedisStreamSource support for cloud provided Redis instances in addition to, cluster local version of redis.
```

**Docs**

### CODE REVIEW

File: source/pkg/adapter/adapter.go [hunk 1]
```diff
@@ -228,9 +236,30 @@ func newPool(address string) *redis.Pool {
 		// Dial is an application supplied function for creating and
 		// configuring a connection.
 		Dial: func() (redis.Conn, error) {
-			c, err := redis.Dial("tcp", address)
-			if err != nil {
-				panic(err.Error())
+			var c redis.Conn
+			if (opt.Password) != "" && a.config.TLSCertificate != "" {
```
Comment (lionelvillard) [inline source/pkg/adapter/adapter.go:240]: nit: (opt.Password) -> opt.Password

File: source/pkg/adapter/adapter.go [hunk 2]
```diff
@@ -228,9 +236,30 @@ func newPool(address string) *redis.Pool {
...
+				roots := x509.NewCertPool()
+				ok := roots.AppendCertsFromPEM([]byte(a.config.TLSCertificate))
+				if !ok {
+					panic(err.Error())
```
Comment (lionelvillard) [inline source/pkg/adapter/adapter.go:244]: panic(err) works too.

File: source/pkg/reconciler/streamsource/controller.go [hunk 3]
```diff
@@ -85,7 +85,17 @@ func NewController(
 			r.updateRedisConfig(ctx, configMap)
 		})
 	} else if !apierrors.IsNotFound(err) {
-		logging.FromContext(ctx).With(zap.Error(err)).Fatal("Error reading ConfigMap'")
+		logging.FromContext(ctx).With(zap.Error(err)).Info("Error reading Redis ConfigMap'")
+	}
+
+	// Get TLS config map and set TLS certificate, to pass data to receive adapter.
+	// Not rolling out new adapters on watch change.
+	if _, err := kubeclient.Get(ctx).CoreV1().ConfigMaps(system.Namespace()).Get(ctx, TLSConfigMapName(), metav1.GetOptions{}); err == nil {
```
Comment (lionelvillard) [inline source/pkg/reconciler/streamsource/controller.go:93]: Usually certificates expire. What's the story for certificate rotation?

File: source/pkg/adapter/adapter.go (hunk 1 above)
Comment (aavarghese) [reply source/pkg/adapter/adapter.go:240]: Done

File: source/pkg/reconciler/streamsource/controller.go (hunk 3 above)
Comment (aavarghese) [reply source/pkg/reconciler/streamsource/controller.go:93]: I think its the same as `config-redis`'s `numConsumers` on-demand update that was requested by asynch team. When a new cert is added to the config-tls CM, we'll need the watch to update the reconciler and create a new redis connection, I assume. 
I can add this to our existing issue for [`config-redis`](https://github.com/knative-sandbox/eventing-redis/issues/8) or create a separate one. That ok? cc: @lionelvillard

File: source/pkg/adapter/adapter.go (hunk 2 above)
Comment (aavarghese) [reply source/pkg/adapter/adapter.go:244]: Done

File: source/pkg/reconciler/streamsource/controller.go (hunk 3 above)
Comment (lionelvillard) [reply source/pkg/reconciler/streamsource/controller.go:93]: sounds good!

**Comment by lionelvillard** [comment]
/lgtm

@aavarghese remove the hold when you are ready.

**Pull request merged**